forecasting/src/autogluon/forecasting/learner/abstract_learner.py

```diff
@@ -0,0 +1,157 @@
+
+        if eval_metric is not None:
+            self.eval_metric = eval_metric
+        else:
+            self.eval_metric = "mean_wQuantileLoss"
+
+        self.is_trainer_present = is_trainer_present
+        if random_seed is None:
+            random_seed = random.randint(0, 1000000)
+        self.random_seed = random_seed
+
+        self.trainer: AbstractTrainer = None
+        self.trainer_type = None
+        self.trainer_path = None
+        self.reset_paths = False
+
+    @property
+    def is_fit(self):
+        return self.trainer_path is not None or self.trainer is not None
+
+    def set_contexts(self, path_context):
+        self.path, self.model_context, self.save_path = self.create_contexts(path_context)
+
+    def create_contexts(self, path_context):
+        model_context = path_context + 'models' + os.path.sep
+        save_path = path_context + self.learner_file_name
+        return path_context, model_context, save_path
+
+    def fit(self, train_data, freq, prediction_length, val_data=None, **kwargs):
+        return self._fit(train_data=train_data,
+                         freq=freq,
+                         prediction_length=prediction_length,
+                         val_data=val_data,
+                         **kwargs)
+
+    def _fit(self, train_data, freq, prediction_length, val_data=None, scheduler_options=None, hyperparameter_tune=False,
+             hyperparameters=None, time_limit=None, use_feat_static_cat=False, use_feat_static_real=False, cardinality=None, **kwargs):
+        raise NotImplementedError
+
+    def refit_full(self, models='all'):
+        return self.load_trainer().refit_full(models=models)
+
+    def predict(self, data, model=None, for_score=False, **kwargs):
+        predict_target = self.load_trainer().predict(data=data, model=model, for_score=for_score, **kwargs)
+        return predict_target
+
+    def evaluate(self, forecasts, tss, **kwargs):
+        quantiles = kwargs.get("quantiles", None)
+        # TODO: difference between evaluate() and score()?
+        if self.eval_metric is not None and self.eval_metric not in ["MASE", "MAPE", "sMAPE", "mean_wQuantileLoss"]:
+            raise ValueError(f"metric { self.eval_metric} is not available yet.")
+
+        # if quantiles are given, use the given on, otherwise use the default
+        if quantiles is not None:
+            evaluator = Evaluator(quantiles=quantiles)
+        else:
+            evaluator = Evaluator()
+
+        num_series = len(tss)
+        agg_metrics, item_metrics = evaluator(iter(tss), iter(forecasts), num_series=num_series)
+        return agg_metrics[self.eval_metric]
+
+    def score(self, data, model=None, quantiles=None):
+        trainer = self.load_trainer()
+        return trainer.score(data, model=model, quantiles=quantiles)
+
+    def leaderboard(self, data=None):
+        trainer = self.load_trainer()
+        return trainer.leaderboard(data)
+
+    def save(self):
```

Comment:
We need to save the trainer and the learner object separately as the trainer cannot be pickled.

Reply:
> I am not sure I understand this save function and why it has to be in this way. Can it be simplified or add some comments on what it does?

We need to save the trainer and the learner object separately as the trainer cannot be pickled.

Reply:
I see, thanks! I find it can be easier to consume in the following, but not strongly. Your call here.
```Python
if self.trainer is not None and self.is_trainer_present:
    self.trainer.save()

trainer = self.trainer
self.trainer = None
save_pkl.save(path=self.save_path, object=self)
self.trainer = trainer
```

Reply:
the logic differs between @huibinshen 's recommendation and the original (`not self.is_trainer_present`) and this version is directly from `tabular`. I'll add a todo and leave as is.